File: src/counter.ts
import * as GenServer from "./gen-server";

interface Opts {
  max?: number;
  name?: string;
}

export default class Counter extends GenServer.t {
  static startLink(opts: Opts = {}): number {
    return GenServer.startLink(Counter, opts, opts);
  }

  static increment(pid: GenServer.Pid, by: number = 1): number {
    return GenServer.call(pid, "increment", by);
  }

  static get(pid: GenServer.Pid): number {
    return GenServer.call(pid, "get");
  }

  static clear(pid: GenServer.Pid): void {
    return GenServer.cast(pid, "clear");
  }

  counter: number;
  max: number;

  constructor({ max }: Opts) {
    super();

    this.counter = 0;
    this.max = max || 10;
  }

  handleCall(action: string, data?: any) {
    const { counter, max } = this;

    switch (action) {
      case "get": {
        return counter;
      }
      case "increment": {
        const newCounter = counter + data;

        if (newCounter <= max) {
          this.counter = newCounter;

          return newCounter;
        } else {
          return counter;
        }
      }
      default: {
        return counter;
      }
    }
  }

  handleCast(action: string) {
    switch (action) {
      case "clear": {
        this.counter = 0;
      }
      default: {
        return this.counter;
      }
    }
  }
}

File: elixir/counter.ex
defmodule Counter do
  use GenServer

  # Client
  def start_link(opts) do
    GenServer.start_link(__MODULE__, [max: opts[:max]], opts)
  end

  def increment(pid) do
    GenServer.call(pid, :increment)
  end

  def get(pid) do
    GenServer.call(pid, :get)
  end

  def reset(pid) do
    GenServer.cast(pid, :reset)
  end

  # Server
  def init(args) do
    max = args[:max] || 10
    %{counter: 0, max: max}
  end

  def handle_call(:increment, _from, %{counter: counter, max: max}) do
    new_counter = counter + 1

    if new_counter <= max do
      {:ok, new_counter, %{counter: new_counter}}
    else
      {:error, :counter_at_max}
    end
  end

  def handle_call(:get, _from, %{counter: counter, max: max}) do
    {:ok, counter, counter}
  end

  def handle_cast(:reset, _from, state) do
    {:ok, %{state | counter: 0}}
  end
end
